File: 300_longest_increasing_subsequence/Solution1/Solution.java
class Solution {
    public int lengthOfLIS(int[] nums) {
        int arrLen = nums.length;
        if (arrLen < 2) return arrLen;

        int[] dp = new int[arrLen];
        dp[0] = 1;
        int result = 1;
        for (int i = 1; i < arrLen; i++) {
            dp[i] = 1;
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j])
                    dp[i] = Math.max(dp[i], dp[j] + 1);
            }
            result = Math.max(result, dp[i]);
        }

        return result;
    }

    public static void main(String[] args) {
        int[] nums = {10, 9, 2, 5, 3, 7, 101, 18};
        System.out.println(new Solution().lengthOfLIS(nums));
    }
}

File: 300_longest_increasing_subsequence/Solution1/Solution.go
package main

import "fmt"

func lengthOfLIS(nums []int) int {
    arrLen := len(nums)
    if arrLen < 2 { return arrLen }

    dp := make([]int, arrLen)
    dp[0] = 1

    for i := 1; i < arrLen; i++ {
        dp[i] = 1
        for j := 0; j < i; j++ {
            if nums[j] < nums[i] {
                dp[i] = max(dp[i], dp[j] + 1)
            }
        }
    }

    result := 0
    for i := 0; i < arrLen; i++ { result = max(result, dp[i]) }

    return result
}

func max(a, b int) int {
    if a > b { return a }
    return b
}

func main() {
    nums := []int{10, 9, 2, 5, 3, 7, 101, 18}
    fmt.Println(lengthOfLIS(nums))
}
